### PR TITLE
feat(sweeper): support heterogeneous P/D hardware [AIC-1772]

### PR DESCRIPTION
**Pull request by jasonqinzhou**
## Why and what changed

Disaggregated P/D search previously applied one `hardware_sku` to both roles. This change adds optional `prefill_hardware_sku` and `decode_hardware_sku` overrides; either omitted role inherits the shared SKU.

Each role is enumerated and KV-qualified against its effective hardware, then paired under the existing total GPU budget. The effective SKUs are preserved in engine arguments, KV-load estimates, performance-model metadata, candidate configurations, and topology provenance. Model, backend, backend version, and GPU budget remain shared.

Two downstream boundaries now fail explicitly:

- Deployment artifact generation rejects heterogeneous P/D because the generator has one global hardware profile. Matching role overrides correctly select their common hardware profile.
- Sweeper rejects a materialized Dynamo Router AIC hook whose system differs from the effective prefill SKU before replay. This includes matching P/D overrides that differ from the fallback. Corrected providers, decode-only overrides, and non-AIC routing remain supported.

The migration guide includes an AIC-to-Sweeper YAML/SDK example and explains the separate unified-CLI and deployment-generator limits.

## Review map

- Risk: medium; start with `sweeper/config.py`, `sweeper/search_space.py`, and `sweeper/search.py`.
- Public/serialized contract: two optional input fields and effective role SKUs in disaggregated output.
- Compatibility: omitted overrides retain shared-hardware behavior; aggregated and AFD hardware behavior is unchanged. No formula or performance-data change.
- Dynamo's provider still needs to consume the effective prefill SKU to support the combinations rejected by the new guard. AISimulate remains independently usable.

## Evidence

Current head: `0f70c6a0f3f3880674895789f121f7414bfcc3fd`, including current `main` at `7dbd110f6459f012d284fdb3080214b6ba204e6e`.

- Rebuilt the native extension with `uv sync --project python/aisimulate --extra dev`.
- `python/aisimulate/.venv/bin/python -m pytest -p no:timeout -q -c pytest.ini tests`: **1009 passed, 5 cross-repository integration skips**.
- `python/aisimulate/.venv/bin/pytest -p no:timeout -q python/aisimulate/tests/unit/generator/`: **347 passed, 4 backend-image skips**.
- Nine new Router boundary cases cover heterogeneous and matching overrides, missing AIC metadata, corrected hooks, decode-only overrides, omitted overrides, non-AIC routing, and aggregated behavior. Provider test module: **34 passed**, including four `Sweeper.run()` submission cases: mismatched hooks never reach `runner.run`, and corrected hooks reach it exactly once.
- Additional materialization probes using unmodified Dynamo Router provider/config source at `4b72f79cb97f144377a5df1ac392b5b72e6dab3e`: both incorrect prefill-SKU cases are rejected; decode-only AIC and non-AIC routing are accepted.
- Ruff check/format passed across all Sweeper source/tests and the changed generator files; `git diff --check` passed.
- [Independent human review](https://github.com/ai-dynamo/aisimulate/pull/186#pullrequestreview-5184255029) on earlier head `f19e892c` records 32 branch snapshots, 3072 paired suggestions, 24 homogeneous native candidate pairs, 24 heterogeneous candidates across three backends, and 12 Bayesian candidates. This is parity/contract evidence, not real-GPU accuracy or convergence qualification.
- Exact-head [Fast CI passed](https://github.com/ai-dynamo/aisimulate/actions/runs/34703261645), and `codeowners` passed.
- NVIDIA bot admission succeeded: trusted `pull-request/186` matches the exact head. [Full CI passed](https://github.com/ai-dynamo/aisimulate/actions/runs/34703515113): 47 successful jobs, six explicitly non-applicable skips, including a successful aggregate `Full CI Success` gate.
- DCO passed after adding the missing sign-off to the older corrective commit. The history repair preserved every commit tree and used an exact-SHA force-with-lease plus a local backup.
- All four actionable review threads are fixed and resolved. GitHub retains the existing human approval. CodeRabbit reviewed the exact final head, assessed minimal merge risk, and identified no remaining merge-blocking issue.

## Modeling and data provenance

Role feasibility and timing identity use repository-native AI Configurator system definitions and performance databases. Existing real-data coverage uses Qwen3-VL-30B FP8 with H200 prefill, GB200 decode, vLLM, a 4096-token sequence bound, and a legal 1+1 GPU pair. Implicit backend versions must agree across both SKUs; otherwise callers must pin a version supported by both.

## Tracking

Related: AIC-1772. Supersedes the role-hardware portion of closed PRs #22 and #23.

### CODE REVIEW

**Comment by jasonqinzhou** [comment]
Codex exact-head self-review finding on `872133854174d40fb4c68c7f40e8393e95b26f93`: the new implicit backend-version agreement guard is behaviorally important but does not yet have a direct regression test for mismatched latest versions. I will add a focused test proving heterogeneous enumeration fails before shape lookup unless a common version is available.

**Comment by jasonqinzhou** [comment]
Addressed the self-review finding in `afe7c2a90e162a4ec8d40b19c30ea4b1a6e41a5d`: the focused test now proves differing implicit latest versions fail before either role begins shape lookup. The full Sweeper suite passes at this head: 497 passed, 5 integration-only skips.

**Comment by jasonqinzhou** [comment]
Codex exact-head self-review completed on `afe7c2a90e162a4ec8d40b19c30ea4b1a6e41a5d`. I rechecked inheritance, per-role feasibility/runtime inputs, shared-version handling, lowering, KV-load capacity, serialized topology/provenance, backward compatibility, and tests. No remaining actionable findings.

**Comment by jasonqinzhou** [comment]
@coderabbitai review

**Comment by jasonqinzhou** [comment]
Validated cross-layer finding from CodeRabbit on `afe7c2a90e162a4ec8d40b19c30ea4b1a6e41a5d`: `from_sweeper_candidate` currently lowers only the shared `hardware_sku`, so it could silently generate homogeneous deployment artifacts from a heterogeneous candidate. The generator contract itself supports only one global hardware profile today, so the narrow safe fix is to reject heterogeneous P/D artifact generation explicitly while preserving heterogeneous Sweeper/replay support. I will add that fail-closed boundary and regression coverage.

**Comment by jasonqinzhou** [comment]
Addressed the validated generator-boundary finding in `7bf27ba4`: heterogeneous Sweeper candidates now fail explicitly at artifact generation instead of silently collapsing both roles to the shared hardware profile. Matching or omitted role SKUs remain compatible. Added generator boundary tests plus an end-to-end H200-prefill/GB200-decode Sweeper test; after merging current `main` at `d11392f3`, generator coverage is 347 passed / 4 skipped and the application suite is 934 passed / 5 skipped.

File: python/aisimulate/src/aiconfigurator/generator/request/sweeper.py [hunk 1]
```diff
@@ -278,6 +288,16 @@ def from_sweeper_candidate(
     backend_version = _required_text(config, "backend_version")
     hardware_sku = _required_text(config, "hardware_sku")
 
+    if mode == "disagg":
+        role_hardware = {
+            role: _role_hardware_sku(config, role, fallback=hardware_sku) for role in ("prefill", "decode")
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply python/aisimulate/src/aiconfigurator/generator/request/sweeper.py:293]: Validated against the current code. The existing matching-role test used the same SKU as `hardware_sku`, so it did not exercise this case. I will set the common resolved role SKU as the generator hardware profile and strengthen the test with a different shared fallback.

[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply python/aisimulate/src/aiconfigurator/generator/request/sweeper.py:293]: Fixed in `5302f0e3`: after confirming the resolved prefill/decode SKUs match, the common value now replaces the shared fallback before `NodeConfig.system_name` and `Platform.hardware_profile` are populated. The regression test now uses `hardware_sku="h200_sxm"` with both role overrides set to `gb200`. Validation: generator 347 passed / 4 skipped; application suite 934 passed / 5 skipped; Ruff check/format passed.

**Comment by jasonqinzhou** [comment]
Validated the latest pre-merge analysis. The fallback `make_candidate_provenance(..., replay_spec=None)` path does carry the new role-SKU fields but lacks a direct regression test, so I am adding one. The separate `aisimulate recommend` input remains intentionally single-hardware; I will document this feature as a Sweeper YAML/SDK contract rather than expanding that distinct CLI in this PR. The modeling-evidence warning does not identify a formula or data change: this patch routes each role through the existing per-SKU KV/performance sources, and the real H200/GB200 integration test plus version-mismatch checks exercise that boundary.

**Comment by jasonqinzhou** [comment]
Addressed the validated fallback-provenance coverage gap in `1f878cde`: added a direct `replay_spec=None` regression asserting both role SKUs in topology, and documented that role overrides are a Sweeper YAML/SDK contract while `aisimulate recommend` retains one shared hardware input. Exact-head application tests: 935 passed / 5 integration-only skips; targeted result tests: 12 passed; Ruff check/format and `git diff --check` passed.

File: docs/sweeper/configuration.md [hunk 2]
```diff
@@ -74,6 +78,13 @@ Each engine role also has lists for `max_num_batched_tokens` and `max_num_seqs`,
 size, GPU-memory-utilization, prefix-caching, and `<role>_forward_model` fields (`op_level` by default,
 or `fpm` for whole-forward timing from a collected FPM cell). A one-item list pins a searched field.
 
+`prefill_hardware_sku` and `decode_hardware_sku` apply only to the ordinary `disagg` branch. Either
+override may be set independently: an omitted role inherits `hardware_sku`. Both roles still share
+the configured model, backend, backend version, and total `gpu_budget`. When `backend_version` is
+omitted, the latest performance-data version for both effective SKUs must match; otherwise pin one
+version supported by both systems. These overrides are part of the Sweeper YAML/SDK contract; the
+separate `aisimulate recommend` input continues to describe one shared hardware SKU.
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply docs/sweeper/configuration.md:86]: Validated against current `ai-dynamo/dynamo` main: `_aic_perf_config_from_candidate` does use `sample["hardware_sku"]` for the disaggregated prefill hook. That consumer is outside this AISimulate-only PR, so I am leaving this thread unresolved and documenting the compatibility blocker rather than hiding it. Core Sweeper KV qualification, backend deployment engine args, and performance metadata already use the effective role SKUs; Router `prefill_load_model.type: aic` requires a Dynamo follow-up to read `prefill_hardware_sku`.

**Comment by jasonqinzhou** [comment]
Codex exact-head self-review completed at `1f878cde13f67b6cd5252f4540f9a8dff0ebac88` against current `main` (`a4c77310`). No additional AISim-local correctness findings. Evidence: application tests 935 passed / 5 integration-only skips; generator tests 347 passed / 4 backend-image skips; Ruff check/format and `git diff --check` passed; exact-head Fast CI passed; CodeRabbit reviewed the exact head. This is review evidence, not approval. Remaining blockers are the unresolved cross-repository Dynamo Router AIC SKU consumer, missing Full CI trusted-copy vetting, DCO on unsigned commit `5302f0e3`, and required human approval.

**Comment by jasonqinzhou** [comment]
Added the requested AIC-to-AISimulate heterogeneous P/D migration guidance in `80f89123`. The guide now maps AIC `prefill_system_name` / `decode_system_name` to the optional Sweeper role-SKU fields, provides a copy-ready H200-prefill/GB200-decode YAML plus SDK invocation, explains one-sided inheritance from `hardware_sku`, and states the shared model/backend/version, unified CLI, deployment-generator, and Dynamo Router AIC boundaries. The embedded YAML validates as `SmartSearchConfig`; exact merged-head application tests are 935 passed / 5 skipped and generator tests are 347 passed / 4 skipped.

**Comment by jasonqinzhou** [comment]
@coderabbitai review

**Comment by jasonqinzhou** [comment]
Validated the exact-head documentation finding. The migration guide now states the Router AIC limitation, but the canonical Sweeper configuration page does not. I will add the same boundary next to the role-SKU semantics so users entering through the Sweeper documentation see it.

**Comment by jasonqinzhou** [comment]
Addressed the documentation consistency finding in signed commit `f19e892c`: the canonical Sweeper configuration page now states that heterogeneous role SKUs are unsupported with Dynamo Router `prefill_load_model.type: aic` until its provider consumes `prefill_hardware_sku`, while shared-SKU behavior is unchanged.

**Comment by jasonqinzhou** [comment]
@coderabbitai review

**Comment by jasonqinzhou** [comment]
Exact-head self-review completed at `f19e892c498c19f5518936580a4cf8b4e7f66bd7`. The AIC migration guide and canonical Sweeper configuration page now consistently document the heterogeneous P/D mapping, fallback semantics, shared model/backend/version constraint, and Router AIC boundary. No remaining AISimulate-local finding. Fast CI and codeowners are green; the PR is still not ready for human review because DCO fails on older unsigned commit `5302f0e3`, Full CI has not started (no NVIDIA trusted-copy branch), human approval is required, and the documented Dynamo Router implementation gap remains external and unresolved.

File: docs/sweeper/configuration.md [hunk 3]
```diff
@@ -74,6 +78,19 @@ Each engine role also has lists for `max_num_batched_tokens` and `max_num_seqs`,
...
+
+The Dynamo Router adapter does not yet support heterogeneous role SKUs with
+`prefill_load_model.type: aic`: its candidate hook uses the shared `hardware_sku` for prefill load
+estimation instead of `prefill_hardware_sku`. Shared-SKU Router AIC behavior is unchanged. Keep
```
Comment (tedzhouhk) [inline docs/sweeper/configuration.md]: **Router AIC follow-up: the limitation also affects matching role overrides.** Reproduced with `hardware_sku: h200_sxm` and both `prefill_hardware_sku` / `decode_hardware_sku` set to `gb200`: the engines correctly use GB200, but enabling `prefill_load_model.type: aic` still produces an H200 prefill load model. The [Dynamo consumer](https://github.com/ai-dynamo/dynamo/blob/0027a8ef0bc108152b1382bfe9584ffa90ec4c7c/components/src/dynamo/router/simulation/provider.py#L85-L102) reads `sample["hardware_sku"]`. The affected condition is effective Prefill SKU differing from the fallback, even when P/D are homogeneous. Please clarify this boundary and have the downstream helper use `sample.get("prefill_hardware_sku") or sample["hardware_sku"]` for disagg, retaining the shared SKU for agg; cover both matching and differing role overrides. This is the existing cross-repository follow-up and does not block my approval of the Sweeper core change.

**Review by tedzhouhk — APPROVED** [review]
Reviewed `f19e892c498c19f5518936580a4cf8b4e7f66bd7` against `68d3e1b727d41cd6bf1336637a8392ed8e4e0ddb`, including native extension compilation from the reviewed source.

Homogeneous compatibility: 32 branch snapshots and 3,072 paired suggestions agree on the legal pool, backend support, structured features, projection, engine arguments, and KV capacity. Matching role overrides, including a different shared fallback, preserve core behavior. Across H200/GB200 and vLLM/SGLang/TensorRT-LLM, 24 controlled native candidate pairs have matching configurations and scores; simulated metrics differ only in a few floating-point means (maximum absolute difference 2.3e-13 ms). The two additive SKU fields and physical execution-time diagnostics were excluded from parity comparison.

Heterogeneous search: both H200-prefill/GB200-decode and the reverse passed across all three backends (24/24 candidates feasible), plus a three-round Bayesian run (12/12 feasible). Additional checks cover role-specific DeepSeek-V3 KV capacity, projector reachability, custom/independent role mappings, version mismatch, and GPU-budget rejection.

Validation: 499 Sweeper tests passed / 5 cross-repository integration skips; 24 generator tests passed; 23 additional boundary probes passed; changed-file Ruff and diff whitespace checks passed.

Approving the AISimulate Sweeper core change. The documented Dynamo Router AIC integration follow-up remains open, including the matching-overrides case described inline. This approval does not establish full CI completion, heterogeneous manifest support, real-GPU predictive accuracy, or optimizer convergence performance.

File: docs/sweeper/configuration.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply docs/sweeper/configuration.md:86]: Confirmed on the current Dynamo provider. I will add an AISimulate candidate-boundary check that rejects a Router AIC hook whose aic_system differs from the effective prefill engine system. This covers heterogeneous roles and matching overrides with a different shared fallback, while accepting providers that already emit the correct system and preserving non-AIC Router modes. Regression tests will exercise materialization and prove mismatched hooks never reach replay.

[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply docs/sweeper/configuration.md:86]: Fixed in 7db90ea2fe1788497294e37bda360bb94a5e4544: Sweeper checks the materialized Dynamo Router AIC hook against the effective prefill SKU before replay. A mismatched or missing AIC system is rejected as an adapter constraint, so it cannot produce an incorrectly scored candidate. Corrected providers, decode-only overrides, shared hardware, and non-AIC load models remain usable. Nine regression cases pass; an additional probe against unmodified Dynamo provider source at 4b72f79cb97f144377a5df1ac392b5b72e6dab3e confirms both bad override cases are rejected and the supported cases are accepted. Full local application suite: 1005 passed / 5 cross-repository skips; generator suite: 347 passed / 4 backend-image skips. The downstream provider enhancement remains useful, but the unsafe combination is now enforced locally.

File: docs/sweeper/configuration.md (hunk 3 above)
Comment (jasonqinzhou) [reply docs/sweeper/configuration.md]: Addressed in 7db90ea2fe1788497294e37bda360bb94a5e4544. Both documentation pages now state that matching role overrides can also differ from the shared fallback. Added a candidate-boundary guard and regression covering hardware_sku=h200_sxm with both roles on gb200: a legacy H200 AIC hook is rejected before replay, while a corrected GB200 hook is accepted. The same rejection was verified against the current unmodified Dynamo provider at 4b72f79cb97f144377a5df1ac392b5b72e6dab3e. This prevents incorrect scoring while preserving the existing provider contract and leaving the downstream enhancement independently deployable.

**Comment by jasonqinzhou** [comment]
@coderabbitai review

**Comment by jasonqinzhou** [comment]
Codex self-review completed on `7db90ea2fe1788497294e37bda360bb94a5e4544` against `main` at `7dbd110f6459f012d284fdb3080214b6ba204e6e`.

Re-read the complete PR diff and the role enumeration, shared-version selection, KV-load capacity, deployment lowering, provider materialization, and provenance consumers. The Router guard inspects the materialized hook, preserves the shared input sample, rejects a wrong or absent AIC prefill system before replay, and accepts a corrected hook. Both the heterogeneous and matching-override cases are covered. No remaining actionable correctness finding.

Validation on this tree: rebuilt native extension; application suite 1005 passed / 5 integration-only skips; generator suite 347 passed / 4 backend-image skips; nine new Router boundary cases; four materialization probes using unmodified Dynamo provider/config source at `4b72f79cb97f144377a5df1ac392b5b72e6dab3e`; Ruff check/format and whitespace checks passed. Existing independent human parity evidence remains applicable to the unchanged role enumeration/modeling code and is linked in the PR description.

Both actionable review threads are fixed and resolved. Exact-head Fast CI and codeowners passed; human approval is retained by GitHub. DCO repair authorization and NVIDIA trusted-runner admission/Full CI are still required before merge. This is technical review evidence, not a replacement for those gates.

File: tests/sweeper/test_search_providers.py [hunk 4]
```diff
@@ -398,6 +398,84 @@ def materialize_replay(self, plan, selection, context):
...
+                supported_backend_topologies=(("*", "*"),),
+                supported_hooks=(HookCapability("dynamo.router", "placement_policy", 1),),
+            )
+
+    selection = {"deployment_mode": mode, "backend": "vllm"}
+    for engine_role in ("prefill", "decode") if mode == "disagg" else ("agg",):
+        selection[f"{engine_role}_max_num_batched_tokens"] = 8192
+        selection[f"{engine_role}_max_num_seqs"] = 256
+    factory = RouterRunnerFactory()
+    prepared, result = search_module._materialize_one(
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply tests/sweeper/test_search_providers.py:450]: Confirmed. The existing direct-helper assertion does not exercise replay submission. I will retain the materialization coverage and add Sweeper.run regression cases for both differing and matching role overrides, with matching runtime-hook capabilities. The test will assert mismatched hooks are reported as adapter constraints and never reach runner.run, while corrected hooks do reach replay.

[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply tests/sweeper/test_search_providers.py:450]: Fixed in 0f70c6a0f3f3880674895789f121f7414bfcc3fd. Four new Sweeper.run cases use matching advertised/runtime capabilities, exercise both differing and matching role overrides, and verify the complete submission path. Mismatched hooks produce one ADAPTER_CONSTRAINT result, no selected candidate, and zero runner.run calls; corrected hooks produce a selected candidate and exactly one runner spec. Existing direct materialization coverage remains. The provider test module now passes all 34 tests, with Ruff check/format and whitespace checks passing. The older missing DCO sign-off was also repaired without changing any file content.

**Comment by jasonqinzhou** [comment]
@coderabbitai review

**Comment by jasonqinzhou** [comment]
Codex self-review completed at `0f70c6a0f3f3880674895789f121f7414bfcc3fd` against `main` at `7dbd110f6459f012d284fdb3080214b6ba204e6e`.

The new test now calls `Sweeper.run()` with matching provider capabilities. Both wrong-system cases are rejected before replay and recorded as adapter constraints; both corrected-system controls reach `runner.run` exactly once. The final provider suite passes all 34 tests, with Ruff check/format and whitespace checks passing. The production files are identical to the previous reviewed and validated head; only four submission-path test cases were added. No remaining actionable correctness finding.

The missing DCO trailer on `5302f0e3` was repaired with an exact-SHA lease and local backup. Each rewritten commit preserves its original file tree; the only older message change is the missing sign-off. DCO now passes. All four review threads are resolved and GitHub retains the existing human approval. Refreshed CodeRabbit review and NVIDIA Full CI admission are the remaining pre-merge validation steps.
